File: eel-core/src/main/scala/io/eels/component/hive/HiveFormat.scala
package io.eels.component.hive

trait HiveFormat {
  def serdeClass: String
  def inputFormatClass: String
  def outputFormatClass: String
}

object HiveFormat {

  def apply(format: String): HiveFormat = format match {
    case "avro" => HiveFormat.Avro
    case "orc" => HiveFormat.Orc
    case "parquet" => HiveFormat.Parquet
    case "text" => HiveFormat.Text
    case other => sys.error("Unknown hive input format: " + other)
  }

  def fromInputFormat(inputFormat: String): HiveFormat = inputFormat match {
    case "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat" => Parquet
    case "org.apache.hadoop.mapred.TextInputFormat" => Text
    case "org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat" => Avro
    case "org.apache.hadoop.hive.ql.io.orc.OrcInputFormat" => Orc
  }

  case object Text extends HiveFormat {
    override def serdeClass: String = "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe"
    override def inputFormatClass: String = "org.apache.hadoop.mapred.TextInputFormat"
    override def outputFormatClass: String = "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"
  }

  case object Parquet extends HiveFormat {
    override def serdeClass: String = "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
    override def inputFormatClass: String = "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat"
    override def outputFormatClass: String = "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat"
  }

  case object Avro extends HiveFormat {
    override def serdeClass: String = "org.apache.hadoop.hive.serde2.avro.AvroSerDe"
    override def inputFormatClass: String = "org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat"
    override def outputFormatClass: String = "org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat"
  }

  case object Orc extends HiveFormat {
    override def serdeClass: String = "org.apache.hadoop.hive.ql.io.orc.OrcSerde"
    override def inputFormatClass: String = "org.apache.hadoop.hive.ql.io.orc.OrcInputFormat"
    override def outputFormatClass: String = "org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat"
  }
}

File: eel-core/src/main/kotlin/io/eels/component/hive/HiveFormat.kt
package io.eels.component.hive

interface HiveFormat {

  fun serdeClass(): String
  fun inputFormatClass(): String
  fun outputFormatClass(): String

  companion object {

    fun apply(format: String): HiveFormat = when (format) {
      "avro" -> HiveFormat.Avro
      "orc" -> HiveFormat.Orc
      "parquet" -> HiveFormat.Parquet
      "text" -> HiveFormat.Text
      else -> throw UnsupportedOperationException("Unknown hive input format $format")
    }

    fun fromInputFormat(inputFormat: String): HiveFormat = when (inputFormat) {
      "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat" -> Parquet
      "org.apache.hadoop.mapred.TextInputFormat" -> Text
      "org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat" -> Avro
      "org.apache.hadoop.hive.ql.io.orc.OrcInputFormat" -> Orc
      else -> throw UnsupportedOperationException("Input format not known $inputFormat")
    }
  }

  object Text : HiveFormat {
    override fun serdeClass(): String = "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe"
    override fun inputFormatClass(): String = "org.apache.hadoop.mapred.TextInputFormat"
    override fun outputFormatClass(): String = "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"
  }

  object Parquet : HiveFormat {
    override fun serdeClass(): String = "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
    override fun inputFormatClass(): String = "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat"
    override fun outputFormatClass(): String = "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat"
  }

  object Avro : HiveFormat {
    override fun serdeClass(): String = "org.apache.hadoop.hive.serde2.avro.AvroSerDe"
    override fun inputFormatClass(): String = "org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat"
    override fun outputFormatClass(): String = "org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat"
  }

  object Orc : HiveFormat {
    override fun serdeClass(): String = "org.apache.hadoop.hive.ql.io.orc.OrcSerde"
    override fun inputFormatClass(): String = "org.apache.hadoop.hive.ql.io.orc.OrcInputFormat"
    override fun outputFormatClass(): String = "org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat"
  }
}